### ISSUE TITLE
A valid Facebook app client token must be set in the AndroidManifest.xml

### ISSUE THREAD

**Issue by mastef** [facebook/facebook-sdk-for-unity#618]
## Checklist

- [x] I've updated to the latest released version of the SDK
- [x] I've searched for existing [GitHub issues](https://github.com/facebook/facebook-sdk-for-unity/issues)
- [x] I've looked for existing answers on [Stack Overflow](https://facebook.stackoverflow.com), the [Facebook Developer Community Forum](https://developers.facebook.com/community/) and the [Facebook Developers Group](https://www.facebook.com/groups/fbdevelopers)
- [x] I've read the [Code of Conduct](CODE_OF_CONDUCT.md)
- [x] This issue is not security related and can safely be disclosed publicly on GitHub

## Environment

Describe your dev environment here, giving as many details as possible. If you have them, make sure to include:

- Unity Editor Version: `2020.3.35f1`
- Unity SDK Version: `13.2.0`
- Installation Platform & Version: `Android` version `any`

## Goals

Init FB SDK on Android

## Expected Results

No error. FB SDK initializing quietly.

## Actual Results

Error thrown : 
<img width="482" alt="image" src="https://user-images.githubusercontent.com/762297/173170832-959ca5fb-0600-4246-b82a-7cebb0b3d258.png">

```java
AndroidJavaException: A valid Facebook app client token must be set in the AndroidManifest.xml or set by calling FacebookSdk.setClientToken before initializing the sdk.

com.facebook.FacebookSdk.sdkInitialize(FacebookSdk.kt:440)
com.facebook.unity.FB.Init(FB.java:103)
com.unity3d.player.UnityPlayer.nativeRender(Native Method)
com.unity3d.player.UnityPlayer.access$300 (Unknown Source: 0)
com.unity3d.player.UnityPlayer$e$1. handleMessage (Unknown Source: 83)
android.os.Handler.dispatchMessage(Handler.java:102)
android.os.Looper. Loop0nce (Looper. java: 226)
android.os.Looper. loop(Looper. java: 313)
com.unity3d.player. UnityPlayer$e.run(Unknown Source: 20)
UnityEngine.AndroidJNISafe.CheckException()
UnityEngine.AndroidJavaObject._CallStatic(String methodName, Object[] args) (at Facebook.Unity.Mobile.Android.AndroidFacebook.Init(String appId, HideUnityDelegate hideUnityDelegate, InitDelegate onInitComplete) 
Facebook.Unity.FB+OnDLLLoaded. Invoke ()
Facebook.Unity.FB+CampiledFacebookLoader.Start ()
```

## Steps to Reproduce

Just configure Facebook SDK in an Unity Project. Set the credentials properly:
<img width="465" alt="image" src="https://user-images.githubusercontent.com/762297/173170920-2308d61d-f9cc-46bd-9ffb-97a020771b14.png">

Generate your AndroidManifest.xml

## Code Samples & Details

1. It looks like the client token is not added to the AndroidManifest. Only `com.facebook.sdk.ApplicationId` is added, but not the `com.facebook.sdk.ClientToken`. Adding it manually to the AndroidManifest fixes it : 
`<meta-data android:name="com.facebook.sdk.ClientToken" android:value="YOUR-CLIENT-TOKEN-HERE" />`

2. Since Client token is marked as (optional) there should be no error in the app if it's missing or the "optional" should be removed and the client token added to the manifest 
<img width="213" alt="image" src="https://user-images.githubusercontent.com/762297/173171092-03a238cc-c452-4314-8786-22c6c583681a.png">

**Comment by snorrsi** [comment]
@mastef It´s required from at least Facebook SDK 13+, should see error regarding that.

**Comment by mastef** [comment]
@snorrsi the issue is rather that a) the Unity Plugin says that it's optional and b) doesn't add it automatically to the AndroidManifest.xml

**Comment by snorrsi** [comment]
@mastef I stand corrected, you are right. Haven't tried Android build yet with it. Maybe the issue name should specify that the cliemt token is not added to the AndroidManifest.xml file

**Comment by mastef** [comment]
@snorrsi the issue name is regarding the error that is thrown in the application. The token not being added is just one part ot the issue. Everything is described in the initial ticket.

**Comment by zakyg** [comment]
The same is happening on iOS. Running the app crashes it on start up (at least in debug builds) with
[Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'Starting with v13 of the SDK, a client token must be embedded in your client code before making Graph API calls.](url)

**Comment by robertf-playstudios** [comment]
I added the client token to the android manifest as outlined in the docs, still getting the error regarding needing to add the token.

**Comment by kerembaydogan** [comment]
@robertf-playstudios

Until it is fixed

for ios :

  ad these lines to `Info.plist`

```
    <key>FacebookClientToken</key>
    <string>YOUR-CLIENT-TOKEN-HERE</string>
```

for android :

  ad this line to `AndroidManifest.xml`

```
<meta-data android:name="com.facebook.sdk.ClientToken" android:value="YOUR-CLIENT-TOKEN-HERE" />
```

**Comment by JarWarren** [comment]
The next release will automatically embed your client tokens into `Info.plist` and `AndroidManifest.xml`. In the meantime, they can be added manually by following @kerembaydogan 's helpful instructions.

**Comment by uchar** [comment]
@JarWarren In the meantime, all of our CI/CD pipelines are broken!
Common! it's 2022, No integration tests? How about running your SDK once on some device before releasing it? your multi-billion company doesn't have any testers?
How about rolling out some hot fix releases and not waiting for months ? Maybe some people are losing time because of it ??

**Comment by vahagn-harutyunyan** [comment]
Also, please do not close issues for which the fixed version is not yet published.

**Comment by JarWarren** [comment]
For clarification, this is not a bug and does not require a new version of the Unity SDK.

In Graph API v13 and up - all calls require a valid Client Token. You can pass this programmatically or, more conveniently, add it to `Info.plist` or `AndroidManifest.xml`.

That said, we know that the Unity SDK has been needing support. More resources are being dedicated to the SDK and we're actively working on a more stable 14.0 release in the near term. We appreciate the candid feedback. Please stay tuned!

**Comment by kerembaydogan** [comment]
@JarWarren With all due respect, this issue is absolutely a bug.

**Comment by uchar** [comment]
@kerembaydogan Yeah,maybe in facebook they call this kind of thing something like Fog instead of Bug, who doesn't love a little mystery while coding?

**Comment by kerembaydogan** [comment]
@JarWarren Also, I believe we should change the status of this issue and reopen it.

Because when something goes wrong and I check the open issues and see an issue that is similar to mine with 16372828+ comments, I feel happy because I am not the only one having that issue.

But if we close that issue without releasing a new working version or a hotfix version, This gives people a false sense of “everything is fine” and wastes their time.

**Comment by JarWarren** [comment]
@kerembaydogan That's fair. I'm reopening this issue until next release. FWIW, the next release removes the "optional" tags and automatically embeds Client Token into the appropriate files as well.

**Comment by kenofori** [comment]
Hi there,
from where we get client token for embedding?

**Comment by kerembaydogan** [comment]
developer.facebook.com > settings > advanced

**Comment by kerembaydogan** [comment]
@kenofori

**Comment by zenlita** [comment]
Hi, thank you for reporting this issue. We would like to update that we have released v14.0.0 that would address this bug. You may download it at https://developers.facebook.com/docs/unity/downloads. If there are further issues, please report bugs at https://developers.facebook.com/support/bugs/. Thank you!

**Comment by noambe** [comment]
I just ran into this issue as well. Spent hours trying to upgrade my FB SDK to something above 11.0.
Nothing worked on my existing Unity 2020 setup. I got a lot of IL2CPP errors when trying to upgrade to V12, V13.2, V14.

Then I moved to the latest Unity 2021, and still fail to upgrade to any version:
**V12:** Applications/2021.3.2f1/Unity.app/Contents/il2cpp/build/deploy/il2cpp --convert-to-cpp --assembly=Library/Bee/artifacts/Android/ManagedStripped/UnityEngine.Purchasing.WinRTStub.dll --assembly=Library/Bee/artifacts/Android/ManagedStripped/Facebook.Unity.dll
**V13.2:** Same issue as in this thread which looked less worse because it's a runtime issue. I'm not quite sure if there is a programmatic way to solve it? **Please let me know if there is**
**V14:** Building Library/Bee/artifacts/Android/d8kzr/libil2cpp.so failed with output:
Library/Bee/artifacts/Android/d8kzr/pxnb_ity.Canvas.o: In function `CanvasJSWrapper_init_m7F37E8D602676BCF848A03B7E02C9FBECE38281F':

So basically I'm stuck on SDK v11.

**Comment by zenith-jacob** [comment]
Try to clean it completely @noambe. Remove everything related to FB SDK, restart Unity, remove again if there's anything left, comment out the FB related code after you remove FB SDK, let it compile, then restart Unity one more time, and then try to install v14. It probably won't work, but worth a shot.

**Comment by noambe** [comment]
> 

@zenith-jacob thanks for reaching out. I just tried that on my original Unity 2020 setup. I Deleted the FacebookSDK folder, commented out all references, restarted Unity, and made sure the game loads.
Then I imported V14, and it again fails in the IL2CPP step:
Exception: Unity.IL2CPP.Building.BuilderFailedException: /Applications/2020.3.24f1/PlaybackEngines/AndroidPlayer/NDK/toolchains/llvm/prebuilt/darwin-x86_64/bin/clang++ @"/var/folders/mn/hy9q4vf566g704xmjmwqlxxm0000gn/T/tmpvYkXj8.tmp" -o

It's probably not related to this issue, but it just means that skipping V13.2 -> V14 isn't a valid solution if V14 doesn't integrate well on either Unity 2020 and 2021.

**Comment by zenith-jacob** [comment]
Honestly I have no idea what that error means, I'm sorry. You'd probably have to compile it with some more verbose logging.
I really advise you to stay out from up-to-date FacebookSDK : >
Why don't you stay on v11?

Also, try to upgrade Unity whenever possible, to 2021 LTS, and then do a clean FB SDK import.
Right now I have a working setup with Unity LTS (2021.3.5f1) and FB SDK 11.0.0 ;)

**Comment by noambe** [comment]
@zenith-jacob good question. The only reason that triggered the upgrade was that the #1 crash I have in Google Play is coming from "Proxy2.onBillingServiceDisconnected". While looking for this online, I came across multiple posts saying it's coming from FB SDK V11: https://stackoverflow.com/questions/69131340/facebook-sdk-11-0-0-cause-crashes-on-android
So I figured it's worth trying to upgrade.

I'm also on a working setup of Unity 2020 + FB SDK 11, and currently prefer not to upgrade to Unity 2021 unless there's a good justification.

**Comment by Legacycz** [comment]
> > 
> 
> @zenith-jacob thanks for reaching out. I just tried that on my original Unity 2020 setup. I Deleted the FacebookSDK folder, commented out all references, restarted Unity, and made sure the game loads. Then I imported V14, and it again fails in the IL2CPP step: Exception: Unity.IL2CPP.Building.BuilderFailedException: /Applications/2020.3.24f1/PlaybackEngines/AndroidPlayer/NDK/toolchains/llvm/prebuilt/darwin-x86_64/bin/clang++ @"/var/folders/mn/hy9q4vf566g704xmjmwqlxxm0000gn/T/tmpvYkXj8.tmp" -o
> 
> It's probably not related to this issue, but it just means that skipping V13.2 -> V14 isn't a valid solution if V14 doesn't integrate well on either Unity 2020 and 2021.

I have the same issue in V14.0 on Unity 2021.3. It is not possible to make the build on Android. So I had to roll back to 13.2 but there is again the problem with client ID.

**Comment by amirhakimnejad** [comment]
Same problem here on building android with SDK v14.
Probably a new issue with the SDK.

**Comment by zenlita** [comment]
> 

Hi @noambe 

This is an indication of mis-configuration of dlls in Unity. When Unity SDK is imported into Unity, unfortunately Unity reconfigures all dlls to 'All Platforms' and this causes an issue when building on a particular platform. We are working on setting up a more detailed guide to unblock developers on common issues such as these. 

So the workaround for now, is to reconfigure the dlls. You will find them in Unity under FacebookSDK -> Plugins and all the dlls are under their various platform folders such as Android / Canvas / Editor / iOS etc. 

Click on them and an inspector will appear (attached screenshot). The example attached shows the configuration for an Android dll. Uncheck All Platforms (if needed) and select Android. Do the same for the rest:

Android.DLL to Android platform
IOS Dll to iOS platform
Canavs to webgl

![Screenshot 2022-07-12 at 9 23 21 AM](https://user-images.githubusercontent.com/31426047/178387725-61119e0e-5696-4d24-9d74-462aea76c40e.png)

**Comment by SnaiperoG3D** [comment]
Small indie studio
